tools/wake-format/formatter.h

```diff
@@ -0,0 +1,320 @@
+
+#include <bitset>
+#include <cassert>
+
+#include "parser/cst.h"
+#include "parser/parser.h"
+
+#define ALWAYS_INLINE inline __attribute__((always_inline))
+
+struct ctx_t {
+  size_t width = 0;
+  size_t nest_level = 0;
+
+  ctx_t nest() {
+    ctx_t copy = *this;
+    copy.nest_level++;
+    return copy;
+  }
+
+  ctx_t sub(const wcl::doc_builder& builder) {
+    ctx_t copy = *this;
+    if (builder.has_newline()) {
+      copy.width = builder.last_width();
+    } else {
+      copy.width += builder.last_width();
+    }
+    return copy;
+  }
+};
+
+inline void space(wcl::doc_builder& builder, uint8_t count) {
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append(" ");
+  }
+}
+
+struct ConsumeWhitespaceAction {
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    while (!node.empty() && (node.id() == TOKEN_WS || node.id() == TOKEN_NL)) {
+      node.nextSiblingElement();
+    }
+  }
+};
+
+struct SpaceAction {
+  uint8_t count;
+  SpaceAction(uint8_t count) : count(count) {}
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    space(builder, count);
+  }
+};
+
+struct NewlineAction {
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    builder.append("\n");
+    space(builder, 4 * ctx.nest_level);
+  }
+};
+
+struct TokenAction {
+  uint8_t token_id;
+  TokenAction(uint8_t token_id) : token_id(token_id) {}
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == token_id);
+    builder.append(node.fragment().segment().str());
+    node.nextSiblingElement();
+  }
+};
+
+struct TokenReplaceAction {
+  uint8_t token_id;
+  const char* str;
+
+  TokenReplaceAction(uint8_t token_id, const char* str) : token_id(token_id), str(str) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == token_id);
+    builder.append(str);
+    node.nextSiblingElement();
+  }
+};
+
+struct WhitespaceTokenAction : public TokenReplaceAction {
+  WhitespaceTokenAction() : TokenReplaceAction(TOKEN_WS, " ") {}
+};
+
+template <class Action1, class Action2>
+struct SeqAction {
+  Action1 action1;
+  Action2 action2;
+  SeqAction(Action1 a1, Action2 a2) : action1(a1), action2(a2) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    action1.run(builder, ctx, node);
+    action2.run(builder, ctx, node);
+  }
+};
+
+template <class F>
+struct WalkAction {
+  F walker;
+  uint8_t node_type;
+
+  WalkAction(F walk, uint8_t node_type) : walker(walk), node_type(node_type) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == node_type);
+    auto doc = walker(ctx.sub(builder), const_cast<const CSTElement&>(node));
+    builder.append(doc);
+    node.nextSiblingElement();
+  }
+};
+
+template <class F, class P>
+struct WalkPredicateAction {
```

Comment:
This is strictly more generic than the above really. I think you can just subsume the above into this one.